syntax = "proto3";

package capsule;

import "capsule/common.proto";

message LoginRequest {
    string token = 1;
}

message LoginConfirmation {
    string token = 1;
    string openid = 2;
    string name = 3;
}

message UserInfo {
    string name = 1;
    repeated bytes aids = 2;
    repeated bytes rids = 3;
}

message ArticleId {
    bytes aid = 1;
}

message RequestId {
    bytes rid = 1;
}

message RequestInfo {
    enum Status {
        done = 0;
        working = 1;
        error = 2;
    }
    Status status = 1;
}

message CreateArticleRequest {
    string name = 1;
}

message AddUrlsRequest {
    bytes aid = 1;
    repeated string urls = 2;
}

message ArticleInfo {
    string title = 1;
    repeated Snapshot snapshots = 2;
}
    
message Query {
    string url = 1;
}

message SnapshotList {
    repeated Snapshot snapshots = 1;
}

message Message {
    enum Type {
        INFO = 0;
        ERROR = 1;
    }

    bytes mid = 1;
    Type type = 2;
    uint64 timestamp = 3;
    string content = 4;
    bool has_read = 5;
}

message MessageList {
    repeated Message messages = 1;
}

service Scheduler {
    rpc TryLogin (LoginRequest) returns (Empty);
    rpc ConfirmLogin (LoginConfirmation) returns (Empty);

    // These interfaces requires token in their context.
    rpc GetUserInfo (Empty) returns (UserInfo);
    rpc GetArticleInfo (ArticleId) returns (ArticleInfo);
    rpc CreateArticle (CreateArticleRequest) returns (ArticleId);
    rpc AddUrlsToArticle (AddUrlsRequest) returns (Empty);
    rpc DeleteArticle (ArticleId) returns (Empty);
    rpc GetRequestInfo (RequestId) returns (RequestInfo);
    rpc GetMessages (Empty) returns (MessageList);

    // These interfaces are open to the public.
    rpc FetchSnapshot (Snapshot) returns (Content);
    rpc ListSnapshot (Query) returns (SnapshotList);

    rpc RegisterWorker (Endpoint) returns (Empty);
    rpc RegisterStorage (Endpoint) returns (Empty);
}

